docs/beefy.md

```diff
@@ -4,97 +4,364 @@ Related issues:
+`best_beefy` block is a power of two. The mental model for round selection is to first finalize
+the mandatory block and then to attempt to pick a block taking into account how fast BEEFY catches
+up with GRANDPA. In case GRANDPA makes progress, but BEEFY seems to be lagging behind, validators
+are changing rounds less often to increase the chance of concluding them.
+
+As mentioned earlier, every time the node picks a new `round_number` (and validator casts a vote) it
+ends the previous one, no matter if finality was reached (i.e. the round concluded) or not. Votes
+for an inactive round should not be propagated.
+
+Note that since BEEFY only votes for GRANDPA-finalized blocks, `session_start` here actually means:
+"the latest session for which the start of is GRANDPA-finalized", i.e. block production might have
+already progressed, but BEEFY needs to first finalize the mandatory block of the older session. See
+[Catch up](#3-catch-up) for more details and also consult [Lean BEEFY](#2-lean-beefy).
+
+In a good networking conditions BEEFY may end up finalizing each and every block (if GRANDPA does
+the same). Practically, with short block times, it's going to be rare and might be excessive, so
+it's suggested for implementations to introduce a `min_delta` parameter which will limit the amount
+of started rounds. The affected component of the formula would be: `best_beefy + MAX(min_delta,
+NEXT_POWER_OF_TWO(...))`, so we start a new round only if the power-of-two component is greater than
+the min delta. Note that if `round_number > best_grandpa` the validators are not expected to start
+any round.
+
+<details>
+<summary>Future: New round selection.</summary>
+
+Better heuristic from Al: in the next beefy round, consider the finalized blocks that
+come after the last beefy-signed block, and select the one whose block number that can be divided
+by 2 the most times.
+(i.e it's forgiving if validators don't agree on neither best beefy block nor best grandpa block)
+
+</details>
+
+### Catch up
+
+Every session is guaranteed to have at least one BEEFY-finalized block. However it also means
+that the round at mandatory block must be concluded even though, a new session has already started
+(i.e. the on-chain component has selected a new validator set and GRANDPA might have already
+finalized the transition). In such case BEEFY must "catch up" the previous sessions and make sure to
+conclude rounds for mandatory blocks. Note that older sessions must obviously be finalized by the
+validator set at that point in time, not the latest/current one.
+
+### Initial Sync
+
+It's all rainbows and unicorns when the node is fully synced with the network. However during cold
+startup it will have hard time determining the current round number. Because of that nodes that
+are not fully synced should not participate in BEEFY protocol at all.
+
+During the sync we should make sure to also fetch BEEFY justifications for all mandatory blocks.
```

Comment:
Why do we need justifications for all mandatory blocks? why not just the last mandatory from the current/previous session?

Reply:
We need to make sure that the validator sets are properly changing, so that BEEFY is kind of self-sufficient, i.e. each BEEFY validator set signs-off it's successor. Without that you would just jump in the middle and assume that the current validator set is indeed legit. While this should be ensured by GRANDPA, we should rather make sure that we don't need this assumption  and also be sure that each full node can serve any BEEFY light client (i.e. contains all mandatory justifications).